### ISSUE TITLE
dynamic invocation on `render!` with CUDA

### ISSUE THREAD

**Issue by ExpandingMan** [dominic-chang/Krang.jl#22]
```julia
function makescene(θs, rmin, rmax)
    mat = Krang.CoordinatePoint()
    meshes = ntuple(Val(3)) do j
        Krang.Mesh(Krang.ConeGeometry(θs, (j, rmin, rmax)), mat)
    end
    Krang.Scene(meshes)
end

function gputest1()
    metric = Krang.Kerr(0.99f0)
    θo = Float32(45 * π / 180)
    sze = 400
    rmin = Krang.horizon(metric)
    rmax = 10.0f0
    ρmax = 15.0f0

    θs = Float32(π/12)
    scene = makescene(θs, rmin, rmax)

    camera = Krang.SlowLightIntensityCamera(metric, θo, -ρmax, ρmax, -ρmax, ρmax, sze, A=CuMatrix)

    store = CUDA.fill(0.0, sze, sze)

    Krang.render!(store, camera, scene)
end
```
On running `gputest1()` I get
```julia
ERROR: InvalidIRError: compiling MethodInstance for Krang.gpu__render!(::KernelAbstractions.CompilerMetadata{…}, ::CuDeviceMatrix{…}, ::CuDeviceMatrix{…}, ::Krang.Mesh{…}) resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to convert)
Stacktrace:
 [1] setindex!
   @ ~/.julia/packages/CUDA/2kjXI/src/device/array.jl:166
 [2] macro expansion
   @ ~/.julia/dev/Krang/src/schemes/schemes.jl:15
 [3] gpu__render!
   @ ~/.julia/packages/KernelAbstractions/xRXlv/src/macros.jl:95
 [4] gpu__render!
   @ ./none:0
Hint: catch this exception as `err` and call `code_typed(err; interactive = true)` to introspect the erronous code with Cthulhu.jl
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, args::LLVM.Module)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/validation.jl:147
  [2] macro expansion
    @ ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:382 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/TimerOutputs/Lw5SP/src/TimerOutput.jl:253 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:381 [inlined]
  [5] emit_llvm(job::GPUCompiler.CompilerJob; toplevel::Bool, libraries::Bool, optimize::Bool, cleanup::Bool, validate::Bool, only_entry::Bool)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/utils.jl:108
  [6] emit_llvm
    @ ~/.julia/packages/GPUCompiler/2CW9L/src/utils.jl:106 [inlined]
  [7] codegen(output::Symbol, job::GPUCompiler.CompilerJob; toplevel::Bool, libraries::Bool, optimize::Bool, cleanup::Bool, validate::Bool, strip::Bool, only_entry::Bool, parent_job::Nothing)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:100
  [8] codegen
    @ ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:82 [inlined]
  [9] compile(target::Symbol, job::GPUCompiler.CompilerJob; kwargs::@Kwargs{})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:79
 [10] compile
    @ ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:74 [inlined]
 [11] #1145
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/compilation.jl:250 [inlined]
 [12] JuliaContext(f::CUDA.var"#1145#1148"{GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}}; kwargs::@Kwargs{})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:34
 [13] JuliaContext(f::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/driver.jl:25
 [14] compile(job::GPUCompiler.CompilerJob)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/compiler/compilation.jl:249
 [15] actual_compilation(cache::Dict{Any, CuFunction}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::typeof(CUDA.compile), linker::typeof(CUDA.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/execution.jl:237
 [16] cached_compilation(cache::Dict{Any, CuFunction}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2CW9L/src/execution.jl:151
 [17] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:380 [inlined]
 [18] macro expansion
    @ ./lock.jl:267 [inlined]
 [19] cufunction(f::typeof(Krang.gpu__render!), tt::Type{Tuple{KernelAbstractions.CompilerMetadata{…}, CuDeviceMatrix{…}, CuDeviceMatrix{…}, Krang.Mesh{…}}}; kwargs::@Kwargs{always_inline::Bool, maxthreads::Nothing})
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:375
 [20] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:112 [inlined]
 [21] (::KernelAbstractions.Kernel{…})(::CuArray{…}, ::Vararg{…}; ndrange::Tuple{…}, workgroupsize::Nothing)
    @ CUDA.CUDAKernels ~/.julia/packages/CUDA/2kjXI/src/CUDAKernels.jl:103
 [22] Kernel
    @ ~/.julia/packages/CUDA/2kjXI/src/CUDAKernels.jl:89 [inlined]
 [23] #29
    @ ~/.julia/dev/Krang/src/schemes/schemes.jl:23 [inlined]
 [24] MappingRF
    @ ./reduce.jl:100 [inlined]
 [25] afoldl(::Base.MappingRF{…}, ::Base._InitialValue, ::Krang.Mesh{…}, ::Krang.Mesh{…}, ::Krang.Mesh{…})
    @ Base ./operators.jl:544
 [26] _foldl_impl
    @ ./reduce.jl:68 [inlined]
 [27] foldl_impl
    @ ./reduce.jl:48 [inlined]
 [28] mapfoldl_impl
    @ ./reduce.jl:44 [inlined]
 [29] mapfoldl
    @ ./reduce.jl:175 [inlined]
 [30] mapreduce
    @ ./reduce.jl:307 [inlined]
 [31] render!
    @ ~/.julia/dev/Krang/src/schemes/schemes.jl:22 [inlined]
 [32] gputest1()
    @ Main ~/src/krang1.jl:50
```

Frustratingly, I cannot find the dynamic invocation as when I try to inspect it with `@code_warntype` and Cthulhu.jl it looks like it should all be static (this is not the first time I've had this problem with GPU).

I have a couple of small fixes incoming that are needed for the camera.

Are you able to test this with CUDA on your end?

**Comment by dominic-chang** [comment]
I don't have CUDA. I've been running everything with Metal. I haven't seen this error with Metal before.

**Comment by dominic-chang** [comment]
Maybe your issue is that you mixed types?
>     store = CUDA.fill(0.0, sze, sze)
 Should this be?
`CUDA.fill(0f0, sze, sze)`

**Comment by ExpandingMan** [comment]
I don't think so, CUDA mostly automatically converts to `Float32`.  I will look into this more thoroughly later and see if I can fix it.

**Update:** Actually you're right, that was making a `Float64` matrix but it didn't really matter for this issue.

**Comment by ExpandingMan** [comment]
Ok, I've figured out what's going on here.  The issue is that I was mistaken by the various different return types of the materials.

First, let me show a working example which is close to my example above but makes no physical sense:
```julia
function makescene(θs, rmin, rmax)
    #mat = Krang.ElectronSynchrotronPowerLawIntensity()
    # CoordinateRadius makes no physical sense here for multiple meshes, but is a working example
    mat = Krang.CoordinateRadius()
    meshes = ntuple(Val(3)) do j
        Krang.Mesh(Krang.ConeGeometry(θs, (j, rmin, rmax)), mat)
    end
    Krang.Scene(meshes)
end

function gputest1()
    metric = Krang.Kerr(0.99f0)
    θo = Float32(45 * π / 180)
    sze = 400
    rmin = Krang.horizon(metric)
    rmax = 10.0f0
    ρmax = 15.0f0

    θs = Float32(π/12)
    scene = makescene(θs, rmin, rmax)

    camera = Krang.SlowLightIntensityCamera(metric, θo, -ρmax, ρmax, -ρmax, ρmax, sze, A=CuMatrix)

    store = CUDA.fill(0.0f0, sze, sze)

    # this *would* work for other material types but only 4 one mesh because of `+`
    #store = cu(Matrix{NTuple{4,Float32}}(undef, sze, sze))

    Krang.render!(store, camera, scene)
end
```

The first problem was that I was confused by the various return types of the materials.  The second problem is that the existing method for `render!` only really works for materials which return scalars, or more specifically for objects which have `+` methods.

The good news is that indeed it's looking like the real math in this package runs just fine on GPU (I haven't done any thorough benchmarking but it also seems like an appreciable speedup over CPU threads).  The bad news is that some of the interface here has issues.

I'm not sure what would be the best way to resolve this.  One possibility is to dispatch the kernels on the material type, so that e.g. for `Intensity` it knows to take the intensity component.  Another, more complicated solution would be to write methods that prepare the output container for `render!` and pick appropriate methods for combining the results.  It seems to me that this package doesn't really handle any kind of occlusion, so maybe another option would be to restrict to one mesh?

Anyway, would be glad to help test whatever you decide on CUDA.  I think it would also be good to add to tests some of the code that would be expected to run on GPU but runs on CPU during testing.  Most of this stuff in this case I think would have been caught by this.

**Comment by dominic-chang** [comment]
Wow great! Glad it works. Yeah, the package doesn't currently handle occlusion. I wasn't going to put it in until I needed to/someone asked me to (I likely will need to in a few weeks/months for my own projects).

 I intended for the default behavior of multiple mesh contributions to be additive blending, but it doesn't make sense to do additive blending on the weird `CoordinatePoint` and `CoordinateRadius` materials I defined. I think I might scrap those coordinate related materials and put their functionality somewhere else since it seems to clash strongly with the philosophy of the code.
